### PR TITLE
Fix Volume widget compatibility with different commands and PipeWire

### PR DESCRIPTION
**Pull request by adolfgatonegro**
Fixes broken functionality of the Volume widget when trying to use commands other than `amixer` (e.g.  #2616) .

This also addresses the incompatibility of the widget with PipeWire (e.g. #2614), possibly making the `PulseVolume` widget unnecessary.

Added two additional configuration parameters, `check_mute_command` and `check_mute_string`, since different audio utilities show the mute status in different ways, and they will fail the hardcoded search for `[off]` in `mixer_out`. This way the user can specify the output that Qtile should look for in their custom command.

### CODE REVIEW

**Comment by elParaguayo** [comment]
Thanks for submitting this. I'll take a look when I get a moment.

File: libqtile/widget/volume.py [hunk 1]
```diff
@@ -185,16 +190,22 @@ def get_volume(self):
...
+            check_mute = subprocess.getoutput(self.check_mute_command)
+
+        if self.check_mute_string in check_mute:
             return -1
 
         volgroups = re_vol.search(mixer_out)
         if volgroups:
             return int(volgroups.groups()[0])
+        elif "pamixer" in self.get_volume_command:
+            return mixer_out
```
Comment (elParaguayo) [inline libqtile/widget/volume.py:208]: I'm not sure about these lines.

Firstly, the method is mean to return an integer but `subprocess.getoutput` returns a string.
Secondly, even if you convert the output to an integer, I suspect that there are other `pamixer` commands that wouldn't give the right output.
Thirdly, what's special about `pamixer` that means we only do this check for that application?

**Comment by elParaguayo** [comment]
These changes should get rid of the test failures.

``` diff
diff --git a/libqtile/widget/volume.py b/libqtile/widget/volume.py
index f198c34e..8b080ace 100644
--- a/libqtile/widget/volume.py
+++ b/libqtile/widget/volume.py
@@ -42,6 +42,7 @@ __all__ = [

 re_vol = re.compile(r"(\d?\d?\d?)%")

+
 class Volume(base._TextBox):
     """Widget that display and change volume

@@ -75,7 +76,7 @@ class Volume(base._TextBox):
         (
             "check_mute_string",
             "[off]",
-            "String to look for to check muted status in the output of ``check_mute_command``."
+            "String to look for to check muted status in the output of ``check_mute_command``.",
         ),
         (
             "step",
```

**Comment by AdolfGatonegro** [comment]
Fixed. Let's see how it goes. 👌

**Comment by AdolfGatonegro** [comment]
After some more testing, the `pamixer` volume commands are parsed correctly by the regex even if the `%` is present. The special check for `pamixer` was not needed after all. These changes work correctly with the default `amixer`, `pamixer`, and `pactl` as intended.

**Comment by AdolfGatonegro** [comment]
Alright, supporting different commands with their various outputs is a bit of a mess. The current implementation should work with any audio tool that outputs the volume level as 1-3 digits plus a percentage sign: `ddd%`. This means that `pamixer` should be used with the `--get-volume-human` instead of just `--get-volume`. 

I think this would be the simplest way to handle it, and let the user know what the widget expects as volume output. Trying to cover for all the idiosyncrasies of the various Linux audio tools would require additional code to check for exceptions and weird edge cases.

File: libqtile/widget/volume.py [hunk 2]
```diff
@@ -185,11 +196,15 @@ def get_volume(self):
             if self.get_volume_command:
                 get_volume_cmd = self.get_volume_command
 
-            mixer_out = self.call_process(get_volume_cmd)
+            mixer_out = subprocess.getoutput(get_volume_cmd)
```
Comment (elParaguayo) [inline libqtile/widget/volume.py:199]: What's the reason for this change?

File: libqtile/widget/volume.py [hunk 3]
```diff
@@ -71,7 +71,18 @@ class Volume(base._TextBox):
...
+            "get_volume_command",
+            None,
+            "Command to get the current volume."
+            "The expected output should include 1-3 numbers and a ``%`` sign.",
+        ),
+        ("check_mute_command", None, "Command to check mute status"),
+        (
+            "check_mute_string",
+            "[off]",
+            "String to look for to check muted status in the output of ``check_mute_command``.",
```
Comment (elParaguayo) [inline libqtile/widget/volume.py:84]: Can you make it clear that, if this string is matched, then the volume is treated as unmuted (assuming I've got that right)?

File: libqtile/widget/volume.py [hunk 4]
```diff
@@ -71,7 +71,18 @@ class Volume(base._TextBox):
         ("volume_app", None, "App to control volume"),
         ("volume_up_command", None, "Volume up command"),
         ("volume_down_command", None, "Volume down command"),
-        ("get_volume_command", None, "Command to get the current volume"),
+        (
+            "get_volume_command",
+            None,
+            "Command to get the current volume."
```
Comment (elParaguayo) [inline libqtile/widget/volume.py:77]: You need a space after the `.` as the next line will be concatenated.

**Review by elParaguayo — CHANGES_REQUESTED** [review]

**Comment by m-col** [comment]
Apologies for this getting forgotten about. I've just tested it and it works as expected, changes LGTM. I've just triggered the tests, hopefully they're all happy.

**Comment by adolfgatonegro** [comment]
No worries, after the last review I kind of forgot to commit the changes elParaguayo requested, and the PR just sat here for a bit. Looks like all the tests except Python 3.11 were successful, not sure what that's about.

File: libqtile/widget/volume.py (hunk 2 above)
Comment (elParaguayo) [reply libqtile/widget/volume.py:202]: Thanks for sharing this.

It's an interesting scenario as `getoutput` doesn't return an error when it's passed a list but it seems that it just runs the first item in the list.

Do you want to submit a fix for this?

Comment (inmaldrerah) [reply libqtile/widget/volume.py:202]: I think just falling back to `self.call_process` would work. By the way, the current version "happens to work", because the output of the command intended to call, by default `amixer -D default sget Master`, is just part of the output of `amixer`, and the first `\d?\d?\d?%` happens to be the volume of `Master`.
